envoy/http/query_params.h

```diff
@@ -15,6 +19,28 @@ namespace Utility {
 using QueryParams = std::map<std::string, std::string>;
 using QueryParamsVector = std::vector<std::pair<std::string, std::string>>;
 
+class QueryParamsMulti {
+private:
+  absl::btree_map<std::string, std::vector<std::string>> data;
+
+public:
+  void remove(absl::string_view key);
+  void add(absl::string_view key, absl::string_view value);
+  void overwrite(absl::string_view key, absl::string_view value);
+  std::string toString();
+  std::string replaceQueryString(const HeaderString& path);
+
+  // This function is only here to allow for sane unit testing, and should not be called outside of
+  // unit tests
+  absl::btree_map<std::string, std::vector<std::string>> copy_map_for_testing() {
```

Comment:
Can we avoid having test specific code in impl? If not, it is probably not a bad idea to just have a getter function. `copy_map_for_testing` is a bit awkward. If this method is needed, we can just call it data(); getData() etc. 
I am not good at naming but a concise function name is preferred IMO